wmfdata/spark.py

```diff
@@ -129,20 +129,27 @@ def create_custom_session(
             category=FutureWarning
         )
 
+    old_session = get_active_session()
+    if old_session:
+        old_session.stop()
+
     if master == "yarn":
         if ship_python_env:
             # The path to our packed conda environment.
             conda_packed_file = conda.pack(**conda_pack_kwargs)
             # This will be used as the unpacked directory name in the YARN working directory.
             conda_packed_name = os.path.splitext(os.path.basename(conda_packed_file))[0]
 
-            # Ship conda_packed_file to each YARN worker.
+            # Ship conda_packed_file to each YARN worker, unless a previous session has done it already.
             conda_spark_archive = f"{conda_packed_file}#{conda_packed_name}"
-            if "spark.yarn.dist.archives" in spark_config:
+            if not old_session and "spark.yarn.dist.archives" in spark_config:
```

Comment:
There will be some cases where the user has made changes to the environment and will want to reship it. Currently, it looks like there's no way to do this without manually running `conda.pack({"force": True})`; even restarting your Jupyter kernel doesn't remove the cached environment.

Should we force a repack every time? It wouldn't be too hard to do that by altering`conda.pack` to add a random string to the filename so the output is different every time. But on the other hand that adds scope and repacking can be a pretty time-consuming operation.

Maybe we should add an argument to `create_custom_session` that forces a repack? Hmm.

Reply:
Why do we need to check `old_session` at all here?  You've  stopped the old one above, so you'll need to do this either way to create the new session, no?

Reply:
Ah wait, yes, this new comment is misleading:
https://github.com/wikimedia/wmfdata-python/blob/699c32e3322e9de38d7ff841a5fba5f731d97ccc/wmfdata/spark.py#L150-L152

What is really happing is that I am not modifying the list at `spark.yarn.dist.archives`, but the archive will be uploaded every time a new Yarn App is spawned, which happens in tandem with creating a new SparkContext. See Yarn Client code [here](https://github.com/apache/spark/blob/7955b3962ac46b89564e0613db7bea98a1478bf2/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L635). 

Will fix comment.

Reply:
> Why do we need to check old_session at all here?

Because we found this weird behavior where the `SparkContext` is gone, but its `SparkConf` seems to persist. So if we set `spark.yarn.dist.archives` previously, we want to avoid setting it again to not get the stack at https://github.com/wikimedia/wmfdata-python/pull/36#issuecomment-1314671723.

Reply:
Fixed misleading `print_err()` on [5daacf3](https://github.com/wikimedia/wmfdata-python/pull/36/commits/5daacf34f65f751035e52f0156b3a6ea510f6009). Also added comment to make the `spark.yarn.dist.archives` hack explicit.

Reply:
> > Why do we need to check old_session at all here?
> 
> Because we found this weird behavior where the `SparkContext` is gone, but its `SparkConf` seems to persist. So if we set `spark.yarn.dist.archives` previously, we want to avoid setting it again to not get the stack at [#36 (comment)](https://github.com/wikimedia/wmfdata-python/pull/36#issuecomment-1314671723).

Further, I found no cleaner way to deal with this as the `spark.yarn.dist.archives` gets set even when I create a `SparkConf` manually. I also tried `SparkConf(loadDefaults=False)`, and that works but then we loose all the good defaults that we have set on our `spark-defaults.conf`.

Reply:
Wow nasty. Hm.  Since we are only trying to work around the problem of SparkConf persisting, and we are only messing with the `spark.yarn.dist.archive`,  instead of checking `old_session`, perhaps we should check if `{conda_spark_archive} not in SparkConf.get("spark.yarn.dist.archives")` ?  That seems to more explicitly address the problem, rather assuming this config is in SparkConf if a session was previously created.

Reply:
Nasty indeed!

Here is the latest version after offline discussion with @ottomata:
https://github.com/wikimedia/wmfdata-python/blob/c95ad645a72d0b85f5b13d7b5a0cd7bfee9145f4/wmfdata/spark.py#L155-L168

Reply:
YES  explicit and very readable, LGTM!